query-engine/query-builders/query-builder/src/lib.rs

```diff
@@ -140,13 +142,55 @@ impl fmt::Display for RelationLink {
 }
 
 #[derive(Debug, Serialize)]
-pub struct DbQuery {
-    pub query: String,
-    pub params: Vec<PrismaValue>,
+#[serde(tag = "type", rename_all = "camelCase")]
+pub enum DbQuery {
+    RawSql {
+        sql: String,
+        params: Vec<PrismaValue>,
+    },
+    TemplateSql {
+        fragments: Vec<Fragment>,
+        placeholder: PlaceholderFormat,
+        params: Vec<PrismaValue>,
+    },
 }
 
 impl DbQuery {
-    pub fn new(query: String, params: Vec<PrismaValue>) -> Self {
-        Self { query, params }
+    pub fn params(&self) -> &Vec<PrismaValue> {
+        match self {
+            DbQuery::RawSql { params, .. } => params,
+            DbQuery::TemplateSql { params, .. } => params,
+        }
+    }
+}
+
+impl fmt::Display for DbQuery {
+    /// Should only be used for debugging, unit testing and playground CLI output.
+    /// The placeholder syntax does not attempt to match any actual SQL flavour.
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        fn format_fragments(f: &mut Formatter<'_>, fragments: &Vec<Fragment>) -> fmt::Result {
+            let mut number = 1;
+            for fragment in fragments {
+                match fragment {
+                    Fragment::StringChunk(s) => {
+                        write!(f, "{}", s)?;
+                    }
+                    Fragment::Parameter => {
+                        write!(f, "${number}")?;
```

Comment:
Let's make use of the `PlaceholderFormat` here

Reply:
Reused `PlaceholderFormat` here, but it ended up in more and harder to read code than the original (more direct) implementation for basically no benefit. Yes, the numbering code is less redundant this way, but it was not worth it, I think.